Custom tasks modeling

IlyasMoutawwakil opened the pull request:
# What does this PR do?

This PR adds custom tasks modeling through a new class `OVModelForCustomTasks` which allows using models that are exported with arbitrary inputs/outputs.

Fixes #611

## how to use
For example to use `google/vit-base-patch16-224` model for `image-classification` while also getting all the `hidden_states`:

- First we need to custom export the model using `main_export` and `model_kwargs={"output_hidden_states": True}` which is passed to the model at inference time, and `custom_export_configs` which is a dictionary of custom export configs that tells the exporter how to handle inputs/outputs (e.g. what to name them).

```python
from transformers import AutoConfig

from optimum.exporters.onnx.model_configs import ViTOnnxConfig
from optimum.exporters.openvino import main_export


custom_task = "hidden_states"  # "attentions"
model_id = "google/vit-base-patch16-224"
base_task = "image-classification"

output = f"vit_with_{custom_task}"
model_kwargs = {f"output_{custom_task}": True}
repo_id = f"IlyasMoutawwakil/vit-with-{custom_task}"


class ViTOnnxConfigWithHiddenStates(ViTOnnxConfig):
    @property
    def outputs(self):
        common_outputs = {}
        common_outputs["logits"] = {0: "batch_size"}

        for i in range(self._normalized_config.num_hidden_layers + 1):
            common_outputs[f"hidden_states.{i}"] = {0: "batch_size"}

        return common_outputs


config = AutoConfig.from_pretrained(model_id)
constom_export_config = ViTOnnxConfigWithHiddenStates(config, task=base_task)


main_export(
    model_name_or_path=output,
    custom_export_configs={"model": constom_export_config},
    model_kwargs=model_kwargs,
    task=base_task,
    output=output,
)
```
- Then we can use `OVModeForCustomTasks`:
```python
import requests
from PIL import Image
from transformers import AutoImageProcessor

from optimum.intel.openvino import OVModelForCustomTasks

image_processor = AutoImageProcessor.from_pretrained(model_id)
url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)
inputs = image_processor(images=image, return_tensors="pt")

ov_model = OVModelForCustomTasks.from_pretrained(output)
ov_outputs = ov_model(**inputs)
print(ov_outputs.keys()) # dict_keys(['logits', 'hidden_states'])

ov_model.push_to_hub(save_directory=output, repository_id=repo_id)
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

IlyasMoutawwakil commented on optimum/intel/openvino/modeling.py:
```diff
@@ -953,3 +954,63 @@ def forward(
+        outputs = self.request(inputs)
+
+        model_outputs = {}
+        for key, value in outputs.items():
+            if len(key.names) == 0:
+                key_names = {"no_name_output_O_o"}
+            else:
+                key_names = key.names
+
+            model_outputs[next(iter(key_names))] = torch.from_numpy(value).to(self.device) if not np_inputs else value
```
will investigate why we get outputs with empty names (@echarlaix any idea ?)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-intel/pr_669). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

IlyasMoutawwakil commented on tests/openvino/test_modeling.py:
```diff
@@ -1525,3 +1526,36 @@ def test_pipeline_image_to_text(self, model_arch: str):
+            ov_outputs = ov_model(**inputs)
+            self.assertIn("logits", ov_outputs)
+            self.assertIsInstance(ov_outputs.logits, TENSOR_ALIAS_TO_TYPE[input_type])
+            # Compare tensor outputs
+            # self.assertTrue(torch.allclose(torch.Tensor(ov_outputs.logits), transformers_outputs.logits, atol=1e-4))
+            # self.assertTrue(
+            #     torch.allclose(torch.Tensor(ov_outputs.attentions), transformers_outputs.attentions, atol=1e-4)
+            # )
```
didn't pass, still investigating why

echarlaix commented on optimum/intel/openvino/modeling.py:
```diff
@@ -953,3 +954,63 @@ def forward(
+        )
+    )
+    def forward(self, **kwargs):
+        np_inputs = isinstance(next(iter(kwargs.values())), np.ndarray)
+
+        inputs = {}
+        for key, value in kwargs.items():
+            inputs[key] = np.array(value) if not np_inputs else value
```
what do you think about something like :
```suggestion
        expected_inputs_names = set(self.input_names)
        inputs_names = set(kwargs)

        if not expected_inputs_names.issubset(inputs_names):
            raise ValueError(
                f"Got unexpected inputs: expecting the following inputs : {', '.join(expected_inputs_names)} but got : {', '.join(inputs_names)}."
            )
        
        np_inputs = isinstance(next(iter(kwargs.values())), np.ndarray)
        inputs = {}
        for input_name in self.input_names:
            inputs[input_name] = np.array(kwargs.pop(input_name)) if not np_inputs else kwargs.pop(input_name)
```

echarlaix reviewed (approved):
Looks great, thanks a lot @IlyasMoutawwakil ! Might be worth adding a test with a customconfig as well, what do you think ?

IlyasMoutawwakil commented on optimum/intel/openvino/modeling.py:
```diff
@@ -953,3 +954,63 @@ def forward(
+        CUSTOM_TASKS_EXAMPLE.format(
+            processor_class=_TOKENIZER_FOR_DOC,
+            model_class="OVModelForCustomTasks",
+            checkpoint="sentence-transformers/all-MiniLM-L6-v2",
```
just pushed an openvino variant to `IlyasMoutawwakil/sbert-all-MiniLM-L6-v2-with-pooler`

IlyasMoutawwakil commented:
@echarlaix I added an openvino variant of `bert-with-pooler` in `IlyasMoutawwakil/sbert-all-MiniLM-L6-v2-with-pooler` and added a custom export test that creates it
